File: DataStructureSwift/DataStructureSwift/HashTable/HashTable.swift
//
//  HashTable.swift
//  DataStructureSwift
//
//  Created by Yubo Qin on 2017/12/19.
//  Copyright © 2017年 Yubo Qin. All rights reserved.
//

import Foundation

class HashTable {
    // inner data structure
    fileprivate var hashArray: [DataItem?]
    // item to mark as deleted
    fileprivate let DELETED = DataItem(key: -1)
    // constructor
    init(initialCapacity: Int) {
        hashArray = [DataItem?](repeating: nil, count: initialCapacity)
    }
    
    fileprivate func hashFunc(_ key: Int) -> Int {
        return key % hashArray.count
    }
    
    func search(_ key: Int) -> Bool {
        var hashVal = hashFunc(key)
        while hashArray[hashVal] != nil {
            if hashArray[hashVal]!.key == key {
                return true
            }
            hashVal += 1
            // 每次都取模防止溢出
            hashVal %= hashArray.count
        }
        return false
    }
    
    func delete(_ key: Int) -> Int {
        var hashVal = hashFunc(key)
        while hashArray[hashVal] != nil {
            if hashArray[hashVal]!.key == key {
                let val = hashArray[hashVal]!.key
                hashArray[hashVal] = DELETED
                return val
            }
            hashVal += 1
            hashVal %= hashArray.count
        }
        return -1
    }
    
    func insert(_ key: Int) {
        let item = DataItem(key: key)
        var hashVal = hashFunc(key)
        while hashArray[hashVal] != nil && hashArray[hashVal]!.key != -1 {
            hashVal += 1
            hashVal %= hashArray.count
        }
        hashArray[hashVal] = item
    }
    
    // inner class of data item
    fileprivate class DataItem {
        fileprivate var key: Int
        init(key: Int) {
            self.key = key
        }
    }
}

File: DataStructurePython/hashtable.py
class HashTable:
    def __init__(self, initial_capacity: int):
        self.hash_array = [None] * initial_capacity
        self.DELETED = self.DataItem(-1)


    def search(self, key: int) -> bool:
        hash_val = self.__hash_func(key)
        while self.hash_array[hash_val]:
            if self.hash_array[hash_val].key == key:
                return True
            hash_val += 1
            # 每次都取模防止溢出
            hash_val %= len(self.hash_array)
        return False


    def delete(self, key: int) -> int:
        hash_val = self.__hash_func(key)
        while self.hash_array[hash_val]:
            if self.hash_array[hash_val].key == key:
                val = self.hash_array[hash_val].key
                self.hash_array[hash_val] = self.DELETED
                return val
            hash_val += 1
            hash_val %= len(self.hash_array)
        return -1


    def insert(self, key: int):
        item = self.DataItem(key)
        hash_val = self.__hash_func(key)
        while self.hash_array[hash_val] and self.hash_array[hash_val].key != -1:
            hash_val += 1
            hash_val %= len(self.hash_array)
        self.hash_array[hash_val] = item
    

    def __hash_func(self, key: int) -> int:
        return key % len(self.hash_array)


    class DataItem:
        def __init__(self, key: int):
            self.key = key